### PR TITLE
Finops

### PR DESCRIPTION
**Pull request by mkuthan**

### CODE REVIEW

File: _posts/2022-04-02-gcp-finops.md [hunk 1]
```diff
@@ -0,0 +1,381 @@
...
+What are the most important cost factors in the Google Cloud Platform for data pipelines in my projects?
+
+* Common resources
+  - vCPU
+  - Memory
+  - Persistent disks
+  - Local disks
+  - Network egress 
+* Dataflow
+  - Streaming service (for real-time pipelines)
```
Comment (piter75) [inline _posts/2022-04-02-gcp-finops.md]:
```suggestion
  - Streaming Engine (for real-time pipelines)
```

File: _posts/2022-04-02-gcp-finops.md [hunk 2]
```diff
@@ -0,0 +1,381 @@
...
+3. Configure the [cloud billing data export](https://cloud.google.com/billing/docs/how-to/export-data-bigquery) to BigQuery
+4. Craft Datastudio report aligned with the developed labeling convention, there is public [example](https://cloud.google.com/billing/docs/how-to/visualize-data) available
+5. Figure out the workarounds for GCP products which do not provide necessary details in the billing data export
+
+I would say that 2) and 5) are the toughest parts of the journey.
+{: .notice--info}
+
+### Labeling convention
+
+There is no silver bullet, the best cloud resources labeling conventions do not exist.
```
Comment (piter75) [inline _posts/2022-04-02-gcp-finops.md]:
```suggestion
There is no silver bullet - perfect cloud resources labeling conventions do not exist.
```

**Review by piter75 — COMMENTED** [review]
Drobiazgi dwa.

Nieźle zebrana w jednym miejscu wiedza budowana przez ostatnie lata ;-)

File: _posts/2022-04-02-gcp-finops.md [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+
+The presented built-in [Cloud Billing Report](https://cloud.google.com/billing/docs/how-to/reports) page provides:
+
+* Spending timeline within daily or monthly data granularity
+* Change since the previous billing period 
+* Ability to filter or split the total costs by projects, services or SKUs
+* Insight into discounts, promotions and negotiated savings
+
+At first, it looks like the cloud billing report plots everything you need, doesn't it?
+Unluckily it has many shortcomings.
```
Comment (fpoon) [inline _posts/2022-04-02-gcp-finops.md]: `Unluckily` sounds a bit weird. Maybe you should use more common `unfortunately` or `alas`, as it won't interrupt reading flow as much

File: _posts/2022-04-02-gcp-finops.md [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+* What if you are using the libraries like Apache Beam or Apache Spark, and the API for setting labels is not publicly exposed?
+* There is also a shared infrastructure like Cloud Composer or Cloud Logging, they also generate significant costs.
+
+Let's tackle all identified challenges one by one.
+
+### Labeling automation
+
+* Use [Terraform](https://www.terraform.io) for "static" cloud resources management like BigQuery datasets, Pubsub topics, Pubsub subscriptions and Cloud storage buckets.
+Prepare the reusable modules with input parameters required to set the mandatory labels.
+If Terraform is too low-level, create some [DSL](https://en.wikipedia.org/wiki/Domain-specific_language) on top.
```
Comment (fpoon) [inline _posts/2022-04-02-gcp-finops.md:275]: I think currently more appropriate suggestion is to use [Terraform CDK](https://github.com/hashicorp/terraform-cdk), as creating your own DSL is a bit of a hassle

**Review by fpoon — CHANGES_REQUESTED** [review]
Nice post! I think you might find [infracost tool](https://www.infracost.io/docs/) interesting, as it might help forecasting infrastructure change costs before even applying them. There is even a [GH action](https://github.com/infracost/actions) for displaying price change in PR review.

That being said, I'm a bit concerned about blurred out prices from reports - It's possible to restore image after gaussian blur, with e. g. [Richardson-Lucy deconvolution algorithm](https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution). I think black bars, although less aesthetically pleasing, are better suited for redaction

File: _posts/2022-04-02-gcp-finops.md (hunk 4 above)
Comment (mkuthan) [reply _posts/2022-04-02-gcp-finops.md:275]: CDK is still very low-level, just another language instead of TF. DSL could be a REAL abstraction.

**Comment by mkuthan** [comment]
> Nice post! I think you might find [infracost tool](https://www.infracost.io/docs/) interesting, as it might help forecasting infrastructure change costs before even applying them. There is even a [GH action](https://github.com/infracost/actions) for displaying price change in PR review.
> 
> That being said, I'm a bit concerned about blurred out prices from reports - It's possible to restore image after gaussian blur, with e. g. [Richardson-Lucy deconvolution algorithm](https://en.wikipedia.org/wiki/Richardson%E2%80%93Lucy_deconvolution). I think black bars, although less aesthetically pleasing, are better suited for redaction

infracost too: Try to think what kind of costs it could predict? And what is a major costs for data pipelines. The tool is not able to forecast anything particular interesting for data pipelines IMHO.

restore blur: there are many easier methods to get real numbers than de-gaussian :) Have you read kevin mitnick?

**Pull request merged**